File: crm/src/main/kotlin/it/polito/crm/entities/Message.kt
package it.polito.crm.entities

import it.polito.crm.dtos.MessageDTO
import it.polito.crm.dtos.MessageInfoDTO
import it.polito.crm.utils.Channel
import it.polito.crm.utils.MessageStatus
import jakarta.persistence.*

@Entity
@Table(name = "messages")
class Message {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val messageId: Long = 0

    @Column(name = "sender")
    lateinit var sender: String

    @Column(name = "date")
    lateinit var date: String

    @Column(name = "subject")
    lateinit var subject: String

    @Column(name = "status")
    var status: MessageStatus = MessageStatus.Received

    @Lob
    @Column(name = "body")
    lateinit var body: String

    @Column(name = "channel")
    lateinit var channel: Channel

    @Column(name = "priority")
    var priority: Int = 0

    @OneToMany(mappedBy = "message")
    val history = mutableSetOf<MessageHistory>()

    fun addNewHistoryRecord(mh: MessageHistory) {
        mh.message = this
        history.add(mh)
    }

    fun toDTO(): MessageDTO =
        MessageDTO(this.messageId, this.sender, this.date, this.subject, this.status, this.body, this.channel, this.priority)

    fun toInfoDTO(): MessageInfoDTO =
        MessageInfoDTO(this.messageId, this.sender, this.date, this.subject, this.status, this.channel, this.priority)
}

File: crm_ui_client/src/api/crm/dto/Message.ts
import {MessageStatus} from "./MessageHistory";

export enum Channel {
    Address, Email, Telephone
}

interface MessageRawData {
    messageId: number | null
    sender: string
    date: string
    subject: string | null
    status: MessageStatus | null
    body: string | null
    channel: Channel
    priority: number | null
    hasAttachments: boolean | null
}

export class Message implements MessageRawData {
    messageId: number | null
    sender: string
    date: string
    subject: string | null
    status: MessageStatus | null
    body: string | null
    channel: Channel
    priority: number | null
    hasAttachments: boolean | null

    constructor(
        messageId: number | null,
        sender: string,
        date: string,
        subject: string | null,
        status: MessageStatus | null,
        body: string | null,
        channel: Channel,
        priority: number | null,
        hasAttachments: boolean | null
    ) {
        this.messageId = messageId
        this.sender = sender
        this.date = date
        this.subject = subject
        this.status = status
        this.body = body
        this.channel = channel
        this.priority = priority
        this.hasAttachments = hasAttachments
    }

    static fromJsonObject(obj: MessageRawData): Message | null {
        try {
            return new Message(
                obj.messageId,
                obj.sender,
                obj.date,
                obj.subject,
                obj.status,
                obj.body,
                obj.channel,
                obj.priority,
                obj.hasAttachments
            )
        } catch (e) {
            console.error(e)
            return null
        }
    }
}
